File: backend/medicine_extractor.js
// medicine_extractor.js

const Tesseract = require('tesseract.js');
const fs = require('fs');
const csv = require('csv-parser');

// Load CSV files
let bucketmap = [];
let bucket = [];
let diseases = [];

fs.createReadStream('/bucketmap.csv')
  .pipe(csv())
  .on('data', (row) => {
    bucketmap.push(row);
  })
  .on('end', () => {
    console.log('bucketmap.csv loaded');
  });

fs.createReadStream('/bucket.csv')
  .pipe(csv())
  .on('data', (row) => {
    bucket.push(row);
  })
  .on('end', () => {
    console.log('bucket.csv loaded');
  });

fs.createReadStream('/dataset_clean1.csv')
  .pipe(csv())
  .on('data', (row) => {
    diseases.push(row);
  })
  .on('end', () => {
    console.log('dataset_clean1.csv loaded');
  });

function drugExtraction(imageFile) {
  return new Promise((resolve, reject) => {
    Tesseract.recognize(
      imageFile,
      'eng',
      {
        logger: (m) => console.log(m),
      }
    )
      .then(({ data: { text } }) => {
        const annotations = extractDrugInformation(text);
        resolve(annotations);
      })
      .catch((error) => reject(error));
  });
}

function extractDrugInformation(text) {
  // Placeholder for extracting drug information
  return {
    text: text,
    entities: [], // Add logic for extracting drug names, dosages, etc.
    bucketmap: bucketmap,
    bucket: bucket,
    diseases: diseases,
  };
}

module.exports = { drugExtraction };

File: backend/medicine_extractor.py
from PIL import Image
import pytesseract
import requests
from io import BytesIO

# If you're using an Azure API or Google Vision API, you can integrate them here.

def drug_extraction(image_file):
    """
    Extract drug-related information from the uploaded image.
    This function uses OCR (pytesseract) to extract text from the image.
    You can replace this with a more sophisticated method (e.g., Azure Cognitive Services).
    """
    # Open the image file
    img = Image.open(image_file)

    # Use pytesseract to extract text from the image
    extracted_text = pytesseract.image_to_string(img)

    # You can further process the extracted text to identify drug names, dosages, etc.
    # For now, we are just returning the extracted text as a placeholder.
    annotations = extract_drug_information(extracted_text)

    return annotations

def extract_drug_information(text):
    """
    Process the extracted text to find drug-related information.
    This is a placeholder function. You can add more sophisticated logic here.
    """
    # For demonstration, we will return the raw text as "annotations".
    # You can use regex or NLP techniques to find specific drug names, dosages, etc.
    annotations = {
        "text": text,
        "entities": []  # Placeholder for any entities like drug names
    }

    # Example: Add logic to find drug names or dosages from the extracted text.
    # You can use regular expressions or NLP models to extract specific information.

    return annotations
